### ISSUE TITLE
Profiler: Show "truly-in-use" memory usage line

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#1739]
This would handle a common pitfall for users when looking at profiles: Users see the memory usage only go up, and assume that the program is leaking memory. In reality the instances that are accumulating are probably invalid, and would be removed by Legion if it actually got squeezed for memory.

IMHO what the user truly wants to know is how much "breathing room" they have on the memory at any given time.

We would need the profiler to be told when instances become valid and invalid, so it can draw a line of the "truly-in-use" memory usage, which only includes valid instances (possibly in addition to the "technically correct" utilization that we show today, where invalid instances are also included).

Assigning @lightsighter to give an estimate on the work, maybe make sub-tasks to assign.

**Comment by lightsighter** [comment]
I was looking at this some yesterday. I realized that there's a bit of a discrepancy between how we currently visualize instance lifetimes and what we probably want to visualized for understanding out of memory conditions. There are actually two different timelines that matter for instances: one at the mapping stage of the pipeline and one for the execution stage of the pipeline.

At the mapping stage of the pipeline, the timeline looks like the following:
1. Legion calls Realm to create the instance (this is the "creation" time of the instance in Realm's profiling)
2. The instance becomes valid when Legion issues copies to make it valid
3. The instance becomes invalid when something else writes to the logical region and maps to a different instance (2 and 3 can repeat an arbitrary number of times)
4. Legion invokes the destroy operation on the instance to make it eligible for deferred allocations

At the execution stage of the pipeline, the timeline looks like the following:
1. The instance is actually granted its allocation and Realm triggers the ready event for the instance (this is the "ready" time for the instance in Realm's profiling) to indicate that the instance can be used.
2. Tasks and copies use the instance
3. The precondition for the instance to be deleted is triggered and Realm reclaims the space for the instance (this is the "destroy" time for the instance in Realm's profiling)

What we render in the profiler today is the execution stage of the pipeline. It shows the actual memory usage by instances during execution so that it aligns with what the tasks and copies are actually doing (e.g., if a task or a copy is running, you can see that the instances that it is using are alive at the same time). This makes sense if all we care about is seeing memory usage by the actual execution.

If we're actually trying to debug OOM conditions though, we don't actually care about the execution stage of the pipeline; instead we care about the mapping stage of the pipeline because we want to see what instances are valid and which have done deferred deletions so that they are eligible for deferred allocations. Unfortunately, you can't just overlay the mapping stage timeline on top of the execution stage of the pipeline because they are almost completely decoupled from each other. For example, if Legion is mapping ahead, you could have the instance be created, go through several cycles of becoming valid/invalid, and then have the destroy method invoked (with a precondition) all before the instance even becomes "ready" during the execution stage of the pipeline. In most cases though the two timelines will overlap but will be kind of unrelated to each other, e.g. the cycles of becoming valid/invalid won't be correlated with when tasks/copies are actually using the instance.

This creates a conundrum: do we want to visualize the mapping stage of instance lifetimes and if so how do we want to render it? I suspect that the answer is that we do want to visualize it, but I also don't think we want to overlay the mapping and execution stages on top of each other as I think it will be confusing. Do we want to have a separate timeline in the profiler for the mapping lifetime of instances? Should it be on by default or maybe should users be able to toggle the memories from the mapping stage to the execution stage and back? Other ideas?

@manopapad @elliottslaughter What do you guys think?

**Comment by lightsighter** [comment]
Right, so perhaps that deserves a different visualization, perhaps one with matplotlib or something that generates a static visualization of all the instances and where they are in memory and how much memory they take up so you can see the holes and the fragmentation and what instances are currently valid (uncollectable). Let's start a separate issue for that since I think we probably do want to keep this issue as well because I do think users will want to see which instances are actually in use and not being used since, as you pointed out, some users get confused by our lazy garbage collection. We might actually be able to do that without any new logging statements, but just using the instance use information that we have for tasks and copies to mark when instances are currently in use.

**Comment by lightsighter** [comment]
> We can't run the profiler after OOM anyway (unless the mapper handles the error and falls back to some other memory, in which case it's not clear what to show), so unless we're fixing that along the way, we wouldn't be able to use Legion Prof anyway.

I have the beginnings of fixing this in the `grandrefactor` branch, although it will be a little bit before Legion shuts down cleanly so we can actually realize this.

> I do think having an execution timeline that is more accurate to what's actually being used would be desirable. E.g., you can take your largest case (that still fits), run it through, and at least observe whether memory usage is growing and what is using what.

I agree with should maintain the current timeline because it accurately aligns with which instances are live at the time when tasks and copies are actually running.

> And execution is relevant here, because even though the decisions happen at mapping time, the lifetimes are ultimately correlated to some tasks. 

This is actually one thing missing from the critical path analysis today, because we don't analyze which instances had to be freed up allow for an allocation to succeed. We could actually technically do this analysis given we know the offsets and sizes of all instances in each memory. We could then align the allocation being ready with when the prior instances using the same space in memory were freed.

>  In contrast, it's not clear if a mapper-based timeline is useful because the mapping timeline is tied to the duration of specific mapper calls (that in general you'd expect to be uncorrelated with actual execution times). 

Right, it's really only relevant at the point that you can't do an allocation and you need to understand what it is that caused that. The history might be relevant in case you wanted to see if fragmentation was growing worse over time, but I agree the timing is definitely decoupled from the execution stage of the pipeline and therefore a bit hard to grasp.

> I also think an OOM-time live dump is potentially also useful.

I think this is the plan in #1797. There will be a mapper runtime call like:

```
void MapperRuntime::dump_memory_state(Memory m, const char *filename);
```

Legion will dump out the mapper runtime state and then there will be a post-processing tool that will parse the log file and either give you text output or present a visualization. You can call that though as many times as you want and make as many dumps as you want so you don't have to wait until OOM to call it, although it will eagerly dump to the file and might be a bit expensive.

**Comment by lightsighter** [comment]
> I realized that there's a bit of a discrepancy between how we currently visualize instance lifetimes and what we probably want to visualized for understanding out of memory conditions. There are actually two different timelines that matter for instances: one at the mapping stage of the pipeline and one for the execution stage of the pipeline.

I was thinking about this some more this week I think I have a plan for how to address this problem for visualization in the profiler. Interestingly I think we can actually do it without changing the logging format; we have all the information we need in the existing profile to do the analysis that we want. What we want to do is visualize the "liveness" of instances during the execution stage of the pipeline (the present) and not the mapping stage (the future). To do this, we leverage all the timing information we have about outstanding tasks, copies, and deppart operations along with knowing which instances each operation is using (something we already record for all operations, except maybe deppart operations, will need to check but easy to add if we don't currently record it). We can define the boundaries of liveness for an instance as follows:

* An instance becomes "live" as soon as there is a pending or executing operation using the instance. 
* An instance becomes "dead" when there are no longer any pending or executing operations using the instance.

So for each instance, we find all the operations that use that instance in the profile and create a time range for each operation from the point where it was created (NOT when it starts executing) to the point where it finishes executing. We then merge all these time ranges together, and that creates the "live" ranges for an instance. The times in between these ranges are when there are no outstanding operations using the instance and the instance could be collected to free up space. We can then compute the utilization on the memory by looking at the liveness ranges for all the instances in that memory and counting up the total space used by the live instances at any point in time. Note in some cases the "liveness" range will start even before the instance is actually allocated (because of deferred allocations) and that is fine, we just don't need to render that part.

One downside to this approach is that it is effectively an approximation of what the mapper might actually be seeing during the mapping stage of the pipeline. There might be something holding a valid reference to an instance that prevents an instance from being collected that is unrelated to having a pending operation that is going to use the instance (e.g., a mapper might have set a `NO_COLLECT` priority on the instance). During our rendering this instance might appear like it is dead in the visualization, but in reality something else is keeping it alive. That said, I feel like this approach is likely to give us a better picture of memory utilization in the profiler than what we have today where it just looks like all the instances are alive all the time.

Thoughts? @manopapad @jjwilke @elliottslaughter

**Comment by lightsighter** [comment]
Actually I thought of a small refinement to liveness that actually mirrors what we do with task waits. We can say:

* An instance becomes "pending live" as soon as there is at least one pending operation that will use the instance
* An instance becomes "actively live" as soon as there is at least one executing operation actively using the instance
* An instance becomes "dead" if there are no pending or executing operations using the instance

So we can track when an instance is "actively" in-use versus when it is live but not currently being used (we'll do this with the 15% coloring like we do with tasks that are ready to resume but cannot yet resume).

**Comment by manopapad** [comment]
Makes sense to me

**Comment by elliottslaughter** [comment]
This seems like a reasonable way to do the visualization, because fundamentally the visual bandwidth is limited and there's only so much you can convey through color shading.

There are still a couple things this doesn't quite capture, if I'm understanding correctly:

 1. Mapper-set GC priorities. I realize that mappers can set this repeatedly, so it's potentially hard to visualize. On the other hand, many mappers effectively set it only once, and so a large percentage (> 80%?) of instances effectively only ever get one GC priority set. If this is `NO_COLLECT` the instance won't ever be collected and the user may be confused why their instances aren't being collected.
    * Even just being able to show GC priority in the instance popup might be sufficient here. I'd be happy with a system where this value "saturates", i.e., it goes from "(unset)" to a specific value to "(varies)" as the mapper sets it 0, 1 and 2+ times. Thus if the mapper is thrashing and setting GC priorities over and over, we won't cut the instance into tiny boxes, but we'll have some idea that the setting is changing.
 2. Validity. As I understand, an instance with valid data will never be collected even if no pending or active tasks are using it. Therefore it isn't really "dead" in the sense of being collectible. I suppose what Mike is saying is that *eventually* either one of two things will happen: either (a) a task will use it, or (b) it will get deleted. However "eventually" can still be a long time and I think this an important dimension to represent, if the top-level task is holding onto regions that it doesn't need for substantial periods of time. I'm not really sure how to deal with this, but I feel like we should at least think about it.

**Comment by lightsighter** [comment]
> Validity. As I understand, an instance with valid data will never be collected even if no pending or active tasks are using it. Therefore it isn't really "dead" in the sense of being collectible.

This is a good point and I think we can actually handle this case too with a profiler analysis! If an instance is not being used, but it is eventually read before it is written again, then (assuming no runtime bugs) we know that it had valid data that whole time and should probably still be considered "live" as well. Now, to do this precisely we would also want to handle the subregion case correctly. For example, let's say an instance can store data for two subregions: r1 and r2. Now let's say the instance is written for the r1 subset by one task, and then later written for the r2 subset by a different task, and lastly read for the r1 subregion by a third task. We still need to detect that it was live for the r1 subregion at the beginning even though another intervening write happened in between because it was to a disjoint subregion. That will require knowing not just which instances operations are using but how they are using them as well as the aliasing information for the different index spaces. We have some of that, but not all of it. Might take a bit more work to make this happen. In the meantime we could do the unsound analysis (write followed by read with no intervening writes) and see how far that gets us.

> Mapper-set GC priorities.

The more general form of this is when valid references are added/removed in the mapping stage of the pipeline. This one is much harder because all these changes are taking place in the "mapper" stage of the pipeline (the future), whereas the profiler is fundamentally rendering the execution stage of the pipeline as we discussed above. At the moment I don't have a good idea of how to propagate the timing of these kinds of state changes back into the execution stage of the pipeline. 

> I'd be happy with a system where this value "saturates", i.e., it goes from "(unset)" to a specific value to "(varies)" as the mapper sets it 0, 1 and 2+ times. Thus if the mapper is thrashing and setting GC priorities over and over, we won't cut the instance into tiny boxes, but we'll have some idea that the setting is changing.

Since this is time-invariant I think we could definitely capture this behavior. As long as it is effectively a summary of what happened over the instance's lifetime and not something where we need to render specific points in time where things change then I think that is not difficult.

**Comment by manopapad** [comment]
Could you just report from Legion when an instance becomes invalid?

**Comment by elliottslaughter** [comment]
> I think we can actually handle this case too with a profiler analysis ...

I think this is all fine, but I agree with @manopapad this is becoming a bit of a heavyweight analysis. Not that we can't do that (now that we have a Rust backend we can do real compute), but if e.g., validity is something Legion already tracks we could just dump that into the logs and avoid the fancy recomputation.

> At the moment I don't have a good idea of how to propagate the timing of these kinds of state changes back into the execution stage of the pipeline.

Am I right that the GC priorities only really matter when the instance is dead for real, i.e., it doesn't contain valid data, and isn't used in any active tasks, but corresponds to a logical region that hasn't been deleted yet?

If so then it seems like the only visualization you'd need would be to show when the mapper is keeping something alive after the last task uses it. I.e., the above conditions are met, but the mapper-assigned priority keeps it alive. Then presumably one of two things happens: either the mapper eventually changes its mind and sets a different priority, or the it doesn't and the instance just sticks around until the corresponding logical region is destroyed.

In either case I think what matters is the "provenance" of the final deletion, and the point in time at which that decision was made. The fact that the matter lives in the "future" isn't really relevant. It really just boils down to (a) what caused the instance to finally be deleted, and (b) when and where that decision was made.

Does that make sense? In other words, I think we can avoid tracking all the intermediate state changes and just visualize that final decision and its timeline.

**Comment by lightsighter** [comment]
> Could you just report from Legion when an instance becomes invalid?
> I think this is all fine, but I agree with @manopapad this is becoming a bit of a heavyweight analysis. Not that we can't do that (now that we have a Rust backend we can do real compute), but if e.g., validity is something Legion already tracks we could just dump that into the logs and avoid the fancy recomputation.

You guys are missing what I said last year. ;) I can indeed just time when instances change validity in the *mapping* stage of the pipeline. That is not meaningful to our rendering in the profiler because the profiler is rendering the execution stage (and only the execution stage currently) of the pipeline. The timings for things happening in the mapping stage have little relevance to how actual execution unfolds. For example, it's not useful to know when an instance becomes invalid in the mapping stage of the pipeline and Legion issues a deferred deletion to Realm. What really matters is when the event precondition for the deferred deletion actually triggers and that instance's memory is reclaimed and given to a new instance so something else can start running. In fact, if you were to try this approach with some of the streaming examples that @reazulhoque has been doing recently, you would see that the timings for ALL of the validity transitions for an instance happen before the instance even exists! The instance often is deferred allocated, becomes valid, transitions to invalid, and is collected for reuse before the timeline in the profiler for the instance even starts because the instance is deferred deleted before the deferred allocation of that same instance is satisfied. In these cases you wouldn't end up rendering anything interesting because the instance would look dead before it was made, which is an apparent paradox that is going to confuse many users.

The reason that this issue has remained open for a year is because of this fundamental impedance mismatch. If it were just as simple as recording some timings and rendering them I would have done it already. The analysis I'm proposing above is a way to render "liveness" of instances in the execution stage of the pipeline that is semantically meaningful. It shows when instances are actually "live" either because they are in-use, about to be used, or have valid data during the execution stage of the pipeline that will be needed later. It's either this or we start new timelines for the "mapping view" of each memory in the profiler which I think will be more confusing to people.

> Am I right that the GC priorities only really matter when the instance is dead for real, i.e., it doesn't contain valid data, and isn't used in any active tasks, but corresponds to a logical region that hasn't been deleted yet?

You're right about the first part, but not the second. GC priorities matter when an instance is no longer valid in the mapping stage of the pipeline at which point it is eligible for collection. There can still be pending users of the instance but that will not stop the collection of the instance; their postcondition events become the precondition to the deferred deletion.

> The fact that the matter lives in the "future" isn't really relevant. It really just boils down to (a) what caused the instance to finally be deleted, and (b) when and where that decision was made.
> Does that make sense? In other words, I think we can avoid tracking all the intermediate state changes and just visualize that final decision and its timeline.

I disagree with this because of the example I gave above. What matters is not when the deferred deletion was issued to Realm, but when the precondition to the deferred deletion is actually triggered and what led to it. Garbage collection takes place in the future and unless we want to start rendering two different timelines (the "future" and "present" views) of memory then we can't just "translate" these timings from the mapping stage of the pipeline to the execution stage of the pipeline.

**Comment by elliottslaughter** [comment]
> ... The timings for things happening in the mapping stage have little relevance to how actual execution unfolds....

I think I'm failing to adequately describe my intentions, because my intention was *not* to use any timing data from the mapping stage. My (intended) suggestion was only that we might use the validity and perhaps the subregion coverage information already computed "for free" in the mapping stage, and log that information while we have it, so that the profiler can then cross-reference this information to avoid bespoke and possibly inaccurate recomputations of Legion's core algorithms later on.

Basically my idea was to go back to this algorithm https://github.com/StanfordLegion/legion/issues/1739#issuecomment-3392355109 (or perhaps your modification https://github.com/StanfordLegion/legion/issues/1739#issuecomment-3392364987) and do that as suggested, with the only change being that we log the "real" validity information out of the mapping stage where they are already computed by Legion. But those additional logging statements *would not* contain any timestamps, because, as mentioned, it's not obvious how that has any relationship to what the user knows and understands anyway. I think logging the task UID and region requirement numbers should be sufficient to correlate the data in the profiler.

**Comment by lightsighter** [comment]
>  with the only change being that we log the "real" validity information out of the mapping stage where they are already computed by Legion. But those additional logging statements would not contain any timestamps, because, as mentioned, it's not obvious how that has any relationship to what the user knows and understands anyway.

There's a fundamental issue with this approach: the ordering that operations map has only a causal but not isomorphic relationship to the order that operations run. That means that the "view" of memory that you have during mapping is not the same as what you have during execution and there is no way to reliably translate the view that you have during mapping to the execution stage. Let me give you two examples:

1. Let's say I have a producer task `t1` that writes a region `r` and then two consumer tasks `t2` and `t3`. They map in order `t1` then `t2` then `t3`. Going off of mapping, `t3` would be the last task to remove a valid reference to the instance and therefore we would conclude that `t3` is the last task using `r` and its execution timing should be used to determine when the instance is no longer live. However, during execution, because `t2` and `t3` are both reading, they can be reordered, and it might be `t2` that runs last and whose timing should be used to determine when the instance for `r` is no longer live.
2. Let's say I have four tasks `t1`, `t2`, `t3`, and `t4` where `t1` writes a region `r1` and `t2` reads `r1` while `t3` writes a region `r2` and `t4` reads `r2`. Further assume that `r1` and `r2` are disjoint. Let's say the tasks map in order and all map to the same instance that can hold both `r1` and `r2`. It might look like the instance becomes invalid between the mapping of `t2` and `t3` (e.g. if `t2` does `READ_DISCARD`), however, when the tasks run, `t1` and `t3` could run first and then `t2` and `t4` run later, in which case the liveness intervals on `r1` and `r2` will overlap even though they didn't during the mapping stage. You can also create the opposite scenario where it looks like the instance is always valid during the mapping stage, but during execution there are two separate liveness ranges. 

There's no way to translate the valid state of an instance and the relative order of its transitions during the mapping stage of the pipeline to the execution stage.

> to avoid bespoke and possibly inaccurate recomputations of Legion's core algorithms later on

To be clear, what I'm proposing adding to the profiler is a liveness analysis which is different than the dependence analysis that Legion does. The inputs are the same, but we're not going to duplicate any logic that exists inside the runtime. It's fundamentally a different algorithm.
